Pr target navigation with language and image input and rnn agent

le-horizon opened the pull request:
Hey Jiangtao, Yingxiang and Wei,

This change utilizes encoding network in tf_agents' actor and value networks to process input image and language.

Only tested with ball as input.  Since other people may also need to process mixed input, pushing this change into master to make it easier for folks to experiment.

Could you take a look?

Thanks,
Le

emailweixu commented on alf/environments/wrappers.py:
```diff
@@ -27,24 +27,43 @@
+
+        raw_space = self.env.observation_space
+        if field_to_stack:
+            assert hasattr(raw_space, 'spaces'), 'observation is not dict'
```
assert isinstance(raw_space, gym.space.Dict)

emailweixu commented on alf/environments/wrappers.py:
```diff
@@ -27,24 +27,43 @@
+        raw_space = self.env.observation_space
+        if field_to_stack:
+            assert hasattr(raw_space, 'spaces'), 'observation is not dict'
+            assert raw_space.spaces[field_to_stack], (
```
If field_to_stack is not in raw_space, I think this will raise exception.
You can use `raw_space.contains(field_to_stack)`

emailweixu commented on alf/environments/wrappers.py:
```diff
@@ -27,24 +27,43 @@
+            assert hasattr(raw_space, 'spaces'), 'observation is not dict'
+            assert raw_space.spaces[field_to_stack], (
+                '{} not in observation.spaces'.format(field_to_stack))
+            space = raw_space.spaces[field_to_stack]
```
space =raw_space[field_to_stack]

emailweixu commented on alf/environments/wrappers.py:
```diff
@@ -53,18 +72,40 @@ def __init__(self, env, stack_size=4, channel_order='channels_last'):
                 stack_size * space.shape[0],
             ] + list(space.shape[1:])
-        self.observation_space = gym.spaces.Box(
+        box_space = gym.spaces.Box(
```
should use the 'high' and 'low' from the original space

le-horizon replied:
changed as recommended.

emailweixu commented on alf/environments/wrappers.py:
```diff
@@ -53,18 +72,40 @@ def __init__(self, env, stack_size=4, channel_order='channels_last'):
+        box_space = gym.spaces.Box(
             low=0, high=255, shape=shape, dtype=np.uint8)
+        if field_to_stack:
+            self.observation_space.spaces[field_to_stack] = box_space
```
self.observation_space does not exist before this line.

emailweixu commented on alf/environments/wrappers.py:
```diff
@@ -27,24 +27,43 @@
+        self._field_to_stack = field_to_stack
+
+        raw_space = self.env.observation_space
+        if field_to_stack:
```
if field_to_stack is not specified, we should stack all items.

le-horizon replied:
Done.

emailweixu commented on alf/environments/wrappers.py:
```diff
@@ -27,24 +27,43 @@
         self._channel_order = channel_order
         self._stack_size = stack_size
+        self._field_to_stack = field_to_stack
+
```
I feel that the code can be simplified by using a list of indices of the spaces that require stacking.

le-horizon replied:
Using a list of the keys now.

emailweixu commented on alf/examples/ac_target_navigation.gin:
```diff
@@ -0,0 +1,87 @@
+get_conv_layers.conv_layer_params=((16, 3, 2), (32, 3, 2))
+conv_layers = @get_conv_layers()
+sentence/generate_network.structure=(
+    'tf.keras.Sequential',
```
Do you think @layers.Sequential, @layers.Embedding are better than string?

emailweixu commented on alf/utils/common.py:
```diff
@@ -714,3 +719,79 @@ def get_action_spec():
+      vocab_size (int): size of the environment's/teacher's vocabulary
+    """
+    assert _env, "set a global env by `set_global_env` before using the function"
+    return _env.observation_spec()['sentence'].shape[0]
```
report error if observation_spec is not correct.

emailweixu commented on alf/utils/common.py:
```diff
@@ -714,3 +719,79 @@ def get_action_spec():
+
+
+@gin.configurable
+def get_conv_layers(conv_layer_params):
```
comments

emailweixu replied:
You can simply use tf_agents.networks.utils.mlp_layers

le-horizon commented:
Dear reviewers,

Tests are passing.  Please take a look and see if there is anything you want me addressed in this PR.

Thanks,
Le

emailweixu commented on alf/utils/common.py:
```diff
@@ -878,6 +893,24 @@ def get_conv_layers(conv_layer_params):
+    """
+    x = tf.keras.Sequential()
+    x.add(
+        tf.keras.layers.Lambda(
```
Why not just return layers.Lambda(...)

emailweixu commented on alf/utils/common.py:
```diff
@@ -878,6 +893,24 @@ def get_conv_layers(conv_layer_params):
+    x = tf.keras.Sequential()
+    x.add(
+        tf.keras.layers.Lambda(
+            lambda x: x + 0, input_shape=input_shape, batch_size=batch_size))
```
why do you need to pass input_shape and batch_size? They are not the required arguments for Lambda()

emailweixu commented on alf/environments/wrappers.py:
```diff
@@ -27,44 +29,144 @@
+                    result[key].append(field)
+            for key in result:
+                if key in self._fields_to_stack:
+                    if (key == 'image'
```
This can be a problem if the key is not "image" but still a image. It's probably better to always respect _channel order

emailweixu commented on alf/environments/wrappers.py:
```diff
@@ -27,44 +29,144 @@
+                        assert isinstance(sp, gym.spaces.MultiDiscrete)
+                        nvec = [
+                            stack_size * sp.nvec[0],
+                        ] + list(sp.nvec[1:])
```
should be: nvec = sp.nvec * stack_size

emailweixu commented on alf/environments/wrappers.py:
```diff
@@ -27,44 +29,144 @@
+                        stacked_space = gym.spaces.MultiDiscrete(nvec)
+
+                    else:
+                        assert name not in self._white_list_fields, (
```
why do we need to have white_list_fields?

emailweixu commented on alf/environments/wrappers.py:
```diff
@@ -27,44 +29,144 @@
+            env,
+            stack_size=4,
+            channel_order='channels_last',
+            fields_to_stack=None,
```
Perhaps you can use the similar way as common.image_scale_transformer to handle the field name to make it more general in the case of nested observation.
https://github.com/HorizonRobotics/alf/blob/d8464e3d16d62309f3e3b64d8a2f57fc21382cbc/alf/utils/common.py#L438

emailweixu commented on alf/environments/utils.py:
```diff
@@ -155,8 +155,8 @@ def create_environment(env_name='CartPole-v0',
         # Each time we can only create one unwrapped env at most
-        py_env = env_load_fn(env_name)
-        py_env.seed(np.random.randint(0, np.iinfo(np.int32).max))
+        py_env = env_load_fn(env_name, wrap_with_process=True)
```
It's fixed by #244

le-horizon replied:
Cool

emailweixu commented on alf/drivers/policy_driver.py:
```diff
@@ -125,17 +125,17 @@ def _training_summary(self, training_info, loss_info, grads_and_vars):
         if self._debug_summaries:
             common.add_action_summaries(training_info.action,
-                                        self.env.action_spec())
+                                        self._env.action_spec())
```
why make this change?

le-horizon replied:
You are right, no need for this.  I misread the code.

le-horizon replied:
This file is a bit confusing, sometimes it uses .env sometimes ._env.  I was trying to make it consistent.

le-horizon commented:
Hey Wei,

Thanks for the suggestions and discussions.

I made several refactors: moved network architecture out of gin file, refactored FrameStack to support nested fields specified by paths.

Please take another look.

Thanks,
Le

emailweixu commented on alf/utils/common.py:
```diff
@@ -864,3 +890,42 @@ def write_gin_configs(root_dir, gin_file):
+
+
+@gin.configurable
+def get_identity_layer():
```
Put it in alf/layers.py. And we make it have same usage as other keras layers.
```python
@gin.configurable
class IdentityLayer(tf.keras.layers.Lambda):
    def __init__(self):
         super().__init__(lambda x: x)
```

emailweixu commented on alf/networks/target_navigation.py:
```diff
@@ -0,0 +1,115 @@
+                    num_state_tiles=None,
+                    num_sentence_tiles=None):
+    """
+    Generate the actor and value networks
```
comment arguments. explain why using tiles

emailweixu commented on alf/environments/wrappers.py:
```diff
@@ -27,44 +29,118 @@
     """Stack previous `stack_size` frames, applied to Gym env."""
 
-    def __init__(self, env, stack_size=4, channel_order='channels_last'):
+    def __init__(
```
As this class is so simple anymore, we need some unittest

emailweixu commented on alf/environments/wrappers_test.py:
```diff
@@ -0,0 +1,75 @@
+        )
+        env.close()
+
+    def test_framestack_some_fields(self):
```
should not use socialbot to test this. It's better to use a fake environment so that it does not depend on socialbot

emailweixu commented on alf/networks/target_navigation.py:
```diff
@@ -0,0 +1,142 @@
+        tf_agents.networks.utils.mlp_layers(
+            conv_layer_params=conv_layer_params,
+            activation_fn=tf.keras.activations.softsign,
+            kernel_initializer=tf.compat.v1.keras.initializers.
```
make these two configurable.

le-horizon replied:
Done.

emailweixu commented on alf/environments/wrappers.py:
```diff
@@ -27,44 +30,161 @@
+                stacked_space = sp
+            return stacked_space
+
+        def _traverse(d, fields_to_stack=None, prefix=""):
```
It seems that this function will change the original space of wrapped the environment. We should make a copy so that the space of the wrapped environment will not be changed.

le-horizon replied:
Good catch!  Done, and this will break the new test added.

emailweixu commented on alf/environments/wrappers.py:
```diff
@@ -27,44 +30,161 @@
+                            fields_to_stack = [
+                                item for item in fields_to_stack
+                                if item != path
+                            ]
```
why need L105-L109?

le-horizon replied:
This is to compute the remaining fields that we haven't stacked, for verification purposes, so that all fields specified in fields_to_stack have to appear in input obs.

emailweixu commented on alf/examples/ac_target_navigation.gin:
```diff
@@ -36,6 +36,9 @@ fc_layer_params=(256,)
 num_sentence_tiles=800
 
+mlp_layers.activation_fn=@tf.keras.activations.softsign
+mlp_layers.kernel_initializer=@tf.compat.v1.keras.initializers.glorot_uniform()
```
It it possible not to use compat.v1 version?

le-horizon replied:
Done.

emailweixu commented on alf/examples/ac_target_navigation.gin:
```diff
@@ -37,7 +37,7 @@ num_state_tiles=800
 
 mlp_layers.activation_fn=@tf.keras.activations.softsign
-mlp_layers.kernel_initializer=@tf.compat.v1.keras.initializers.glorot_uniform()
+mlp_layers.kernel_initializer=@tf.keras.initializers.GlorotUniform
```
Missing ()?: GlorotUniform()

le-horizon replied:
You are right!  Fixed.  This is scary, the code can still run without complaint..
